librosa/core/pitch.py

```diff
@@ -336,3 +339,278 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Autocorrelation.
+    y_frames = util.frame(y, frame_length=2*win_length, hop_length=hop_length)
+    n_frames = y_frames.shape[-1]
+    a = np.fft.fft(y_frames, 2*win_length, axis=0)
+    b = np.fft.fft(y_frames[win_length::-1, :], 2*win_length, axis=0)
+    acf_frames = np.fft.ifft(a * b, 2*win_length, axis=0).real[win_length:]
+
+    # Energy terms.
+    window = scipy.signal.windows.boxcar(win_length + 1)
+    energy_frames = scipy.signal.fftconvolve(
+        y_frames**2, window[:, None], mode="valid", axes=0)
+
+    # Difference function.
+    yin_frames = energy_frames[0, :] + energy_frames - 2*acf_frames
+
+    # Cumulative mean normalized difference function.
+    min_period = np.maximum(int(np.floor(sr / fmax)), 2)
+    max_period = np.minimum(int(np.ceil(sr / fmin)), win_length - 1)
+    yin_numerator = yin_frames[min_period - 1:max_period + 1, :]
+    tau_range = np.arange(1, max_period + 1)[:, None]
+    cumulative_mean = np.cumsum(yin_frames[1:max_period + 1, :], axis=0) / tau_range
+    yin_denominator = cumulative_mean[min_period - 2:max_period, :]
+    yin_frames = yin_numerator / (yin_denominator + util.tiny(yin_denominator))
+
+    # Parabolic interpolation.
+    parabola_a = (yin_frames[:-2, :] + yin_frames[2:, :] - 2*yin_frames[1:-1, :]) / 2
+    parabola_b = (yin_frames[2:, :] - yin_frames[:-2, :]) / 2
+    parabolic_shifts = -parabola_b / (2*parabola_a + util.tiny(parabola_a))
+    parabolic_shifts[np.abs(parabolic_shifts) > 1] = 0
+
+    # Find Yin candidates and probabilities.
+    # The implementation here follows the official pYIN software which
+    # differs from the method described in the paper.
+    # 1. Define the prior over the thresholds.
+    peak_thresholds = np.linspace(0, 1, n_thresholds + 1)
+    beta_sf = scipy.stats.beta.sf(peak_thresholds, 2, 18)
+
+    # 2. Find local minima in the difference function.
+    is_trough = np.logical_and(
+        yin_frames[1:-1, :] < yin_frames[:-2, :],
+        yin_frames[1:-1, :] < yin_frames[2:, :])
+    yin_frames = yin_frames[1:-1, :]
+
+    # 3. Find minima below each threshold.
+    is_threshold_trough = np.logical_and(
+        is_trough[:, :, None],
+        yin_frames[:, :, None] < peak_thresholds[None, None, 1:])
+
+    # 4. Count number of thresholds above each minima.
+    global_min = np.argmin(yin_frames, axis=0)
+    count_above = n_thresholds - np.count_nonzero(is_threshold_trough, axis=-1)
+    count_above_at_min = count_above[global_min, range(n_frames)]
+
+    # 5. Compute probability of each minima and normalize.
+    yin_probs = beta_sf[count_above]
+    sum_probs = np.sum(yin_probs, axis=0)
+    sum_probs[sum_probs < util.tiny(sum_probs)] = 1
+    yin_probs *= yin_probs[global_min, range(n_frames)] / sum_probs
+
+    # 6. Add to global minimum when all minima are above the thresholds.
+    yin_probs[global_min, range(n_frames)] += 0.01*(1 - beta_sf[count_above_at_min])
```

Comment:
do these constants `0.01` and `0.99` depend upon the chosen number of thresholds?

Reply:
The `0.01` constant is based on the `a` term in equation (4) of the pYIN paper. It doesn't depend on the number of thresholds.

Reply:
ok, great. Please write a comment saying that this constant arises from Equation (4).